unlock-app/src/utils/Error.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * This module provides types and constructors for error values.
+ * Example usage:
+ *
+ * import { StorageError } from '/path/to/Error.ts'
+ * // oh no, an unrecoverable error has occurred!
+ * const err = StorageError.Fatal('we ran out of magnetic tape.')
+ * dispatch(setError(err))
+ */
+
+type ErrorLevel = 'Fatal' | 'Warning'
+
+type ErrorKind =
+  | 'ApplicationError'
+  | 'FormValidationError'
+  | 'LogInError'
+  | 'SignUpError'
+  | 'SignatureError'
+  | 'StorageError'
+  | 'TransactionError'
+
+interface UnlockError<L extends ErrorLevel, K extends ErrorKind> {
+  level: L
+  kind: K
+  message: string
+}
+
+// We very purposely do not export these: error messaging should be centralized
+// here. We want to avoid ad-hoc creation of unhandled error types.
+function fatalMaker<K extends ErrorKind>(kind: K) {
+  return (message: string): UnlockError<'Fatal', K> => ({
+    level: 'Fatal',
+    kind,
+    message,
+  })
+}
+
+function warningMaker<K extends ErrorKind>(kind: K) {
+  return (message: string): UnlockError<'Warning', K> => ({
+    level: 'Warning',
+    kind,
+    message,
+  })
+}
+
+// This is here to enforce that the Fatal property of a group of error
+// constructors will actually be fatal -- no accidentally ignored errors.
+interface ErrorMakers<K extends ErrorKind> {
```

Comment:
This is kind of nice, it constrains what can live in one of these objects, so you can't accidentally have 2 functions that produce warnings or accidentally put the wrong kind of error in the container. So we can be confident that `StorageError.Fatal` can only ever create a fatal error for the storage service.